### DISCUSSION TITLE
Collecting data about primary control (CIO)?

### DISCUSSION THREAD

**Discussion by rareman99** [n27051538/solaris_exporter#11] [Ideas]
look like no metrics collecting primary or IO controls? Cloud you help to add this metrics such as command “ldm ls” etc.

**Comment by n27051538** [comment]
Dear @rareman99 ,
I have added LdomsLsCollector in version 'v2022May04' with 'ldm ls' command results for you.
Note that monitor user needs new auth '**solaris.ldoms.read**'. This auth is added to updated script [solaris_exporter_role.sh](https://github.com/n27051538/solaris_exporter/blob/master/solaris_exporter_role.sh). 

New metrics:

        # HELP solaris_exporter_ldoms ldoms counters
        # TYPE solaris_exporter_ldoms gauge
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="ncpu"} 32.0
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="mem"} 3.4359738368e+010
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="util"} 0.5
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="uptime_seconds"} 659559.0
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="norm_util"} 0.4
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="state"} 0.0
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="flags"} 10110.0
        solaris_exporter_ldoms{host="host01",ldom="primary",statistic="console_port"} -1.0
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="ncpu"} 48.0
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="mem"} 2.147483648e+011
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="util"} 0.7
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="uptime_seconds"} 659576.0
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="norm_util"} 0.6
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="state"} 0.0
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="flags"} 10000.0
        solaris_exporter_ldoms{host="host01",ldom="dom02",statistic="console_port"} 5002.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="ncpu"} 48.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="mem"} 2.147483648e+011
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="util"} -1.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="uptime_seconds"} -1.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="norm_util"} -1.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="state"} 2.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="flags"} 0.0
        solaris_exporter_ldoms{host="host01",ldom="dom01",statistic="console_port"} 0.0
        # HELP solaris_exporter_ldom_collector_timeouts_total Number of times when ldom collector ran more than 3 seconds
        # TYPE solaris_exporter_ldom_collector_timeouts_total counter
        solaris_exporter_ldom_collector_timeouts_total 0.0
        # HELP solaris_exporter_ldom_collector_timeouts_created Number of times when ldom collector ran more than 3 seconds
        # TYPE solaris_exporter_ldom_collector_timeouts_created gauge
        solaris_exporter_ldom_collector_timeouts_created 1.651672718173066e+09
        # HELP solaris_exporter_ldom_collector_errors_total Number of times when ldom collector ran with errors
        # TYPE solaris_exporter_ldom_collector_errors_total counter
        solaris_exporter_ldom_collector_errors_total 0.0
        # HELP solaris_exporter_ldom_collector_errors_created Number of times when ldom collector ran with errors
        # TYPE solaris_exporter_ldom_collector_errors_created gauge
        solaris_exporter_ldom_collector_errors_created 1.651672718173104e+09
        # HELP solaris_exporter_ldom_collector_processing Time spent processing request
        # TYPE solaris_exporter_ldom_collector_processing gauge
        solaris_exporter_ldom_collector_processing 0.10866618156433105

Also note, that you will need to review source code to look meanings of codes of metrics
 
            solaris_exporter_ldoms{statistic="state"} 0.0
            solaris_exporter_ldoms{statistic="flags"} 10110.0

Those are listed [here](https://github.com/n27051538/solaris_exporter/blob/3d613107e221199d71c8fb19fc7eb4422783d18f/solaris_exporter.py#L1102) and [here](https://github.com/n27051538/solaris_exporter/blob/3d613107e221199d71c8fb19fc7eb4422783d18f/solaris_exporter.py#L1118).

**Comment by rareman99** [comment]
Thank you very much. If possible to add solaris cluster (v4.0) to the metrics :)

**Reply by n27051538** [reply]
I haven't touch any solaris clusters for about 10 years and have no any cluster alive.
You could write this collector yourself.
Or you could write some shell script which will produce some metrics file via cron job.
The output of your script needs to be placed into file with name **/opt/solaris_exporter/*.prom**.
File will be parsed by [this](https://github.com/n27051538/solaris_exporter/blob/c64c0911106530d7139746d3f373832825377ae0/solaris_exporter.py#L1047) code that was suggested by [Marcel Peter](mailto:Marcel.Peter1@swisscom.com). I haven't tested this collector still.

**Reply by rareman99** [reply]
Thanks
